### PR TITLE
EG-2294 Suspend Membership flow implementation

### PR DESCRIPTION
**Pull request by tubibuto**
## Description

This is the forth PR in the series of membership management flow PRs. We build on top of [Activate Membership flow PR](https://github.com/corda/corda/pull/6369) to introduce flow for suspension of membership inside Business Network.

First we implement `SuspendMembershipFlow` which takes membershipId which linear ID of the MembershipState we want to suspend. This is the workflow:

- Check initiator is authorised to suspend membership
- Initiator builds membership suspension transaction and sends it for signing to all active members authorised to modify membership
- Transaction is finalised and sent to all observers who are initiator, suspended member and all members authorised to modify membership
- If the member is first time onboarded (his membership state was in pending status before suspension), then all the authorised memberships are sent to it. It will also get all non revoked memberships if it's authorised to modify memberships.

In the end we add test coverage inside `SuspendMembershipFlowTest` and amend `MembershipFlowTest` to support more test helper methods.

## Test Plan

Ensure all existing and new Business Network related tests pass.

Deploy nodes, create Business Network, suspend membership and query parties vaults for sanity check.

Membership suspension and vault query (BNO node):
Screenshot 2020-06-18 at 12 35 47

Vault query (Member1 node) - notice it also picked up authorised member state (BNO):
Screenshot 2020-06-18 at 12 35 56

### CODE REVIEW

**Review by bpaunescu — APPROVED** [review]
lgTM

File: business-networks/workflows/src/main/kotlin/net/corda/bn/flows/SuspendMembershipFlow.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ */
+@InitiatingFlow
+@StartableByRPC
+class SuspendMembershipFlow(private val membershipId: UniqueIdentifier) : MembershipManagementFlow<SignedTransaction>() {
+
+    @Suspendable
+    override fun call(): SignedTransaction {
+        val databaseService = serviceHub.cordaService(DatabaseService::class.java)
+        val membership = databaseService.getMembership(membershipId)
+                ?: throw FlowException("Membership state with $membershipId linear ID doesn't exist")
```
Comment (alexandradh) [inline business-networks/workflows/src/main/kotlin/net/corda/bn/flows/SuspendMembershipFlow.kt]: An `IllegalArgumentException` or a dedicated exception type would be more fitting.

File: business-networks/workflows/src/main/kotlin/net/corda/bn/flows/SuspendMembershipFlow.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                .addInputState(membership)
+                .addOutputState(outputMembership)
+                .addCommand(MembershipContract.Commands.Suspend(), signers.map { it.owningKey })
+        builder.verify(serviceHub)
+
+        // send info to observers whether they need to sign the transaction
+        val observerSessions = observers.map { initiateFlow(it) }
+        val finalisedTransaction = collectSignaturesAndFinaliseTransaction(builder, observerSessions, signers)
+
+        // send authorised memberships to new activated or new suspended member (status moved from PENDING to ACTIVE/SUSPENDED)
```
Comment (alexandradh) [inline business-networks/workflows/src/main/kotlin/net/corda/bn/flows/SuspendMembershipFlow.kt]: Just to new suspended member, and status moving to SUSPENDED.

**Review by alexandradh — CHANGES_REQUESTED** [review]
Please see my comments inline.

**Review by alexandradh — APPROVED** [review]
LGTM

**Review by alexandradh — APPROVED** [review]

**Pull request merged**